From 816ea9be4b8b53d96b4123daed0291c351f610cf Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Tue, 22 Oct 2024 18:00:07 +0200
Subject: [PATCH 1468/1695] plat-stm32mp1: remove shared_resource support for
 GPIOs

Remove management of GPIO secure state through the shared_resources.c
driver since it is now handled straight through the firewall framework.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 2848074a7368cddbc2ec19d5fd382eea02504a0f plat-stm32mp1: remove shared_resource support for GPIOs]
Change-Id: I7cbde3163f890a6ca116077db161b1675c712756
Link: https://github.com/OP-TEE/optee_os/commit/40848ef1888329366204fdbfd047b50d762c0635
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/413527
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 .../arch/arm/plat-stm32mp1/shared_resources.c | 143 +-----------------
 core/arch/arm/plat-stm32mp1/stm32_util.h      |  71 ---------
 core/drivers/stm32_gpio.c                     |   3 -
 3 files changed, 1 insertion(+), 216 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/shared_resources.c b/core/arch/arm/plat-stm32mp1/shared_resources.c
index b527e4786..a5564760b 100644
--- a/core/arch/arm/plat-stm32mp1/shared_resources.c
+++ b/core/arch/arm/plat-stm32mp1/shared_resources.c
@@ -132,25 +132,6 @@ static __maybe_unused const char *shres2str_state(enum shres_state id)
 	return shres2str_state_tbl[id];
 }
 
-/* GPIOZ bank pin count depends on SoC variants */
-/* A light count routine for unpaged context to not depend on DTB support */
-static int gpioz_nbpin = -1;
-
-static unsigned int get_gpioz_nbpin(void)
-{
-	if (gpioz_nbpin < 0)
-		panic();
-
-	return gpioz_nbpin;
-}
-
-void stm32mp_register_gpioz_pin_count(size_t count)
-{
-	assert(gpioz_nbpin == -1);
-
-	gpioz_nbpin = count;
-}
-
 static void register_periph(enum stm32mp_shres id, enum shres_state state)
 {
 	assert(id < STM32MP1_SHRES_COUNT &&
@@ -181,11 +162,7 @@ static void register_periph(enum stm32mp_shres id, enum shres_state state)
 	case STM32MP1_SHRES_GPIOZ(5):
 	case STM32MP1_SHRES_GPIOZ(6):
 	case STM32MP1_SHRES_GPIOZ(7):
-		if ((id - STM32MP1_SHRES_GPIOZ(0)) >= get_gpioz_nbpin()) {
-			EMSG("Invalid GPIO %u >= %u",
-			     id - STM32MP1_SHRES_GPIOZ(0), get_gpioz_nbpin());
-			panic();
-		}
+		panic("Deprecated registering of GPIOz resources");
 		break;
 	default:
 		break;
@@ -196,16 +173,6 @@ static void register_periph(enum stm32mp_shres id, enum shres_state state)
 	/* Explore clock tree to lock secure clock dependencies */
 	if (state == SHRES_SECURE) {
 		switch (id) {
-		case STM32MP1_SHRES_GPIOZ(0):
-		case STM32MP1_SHRES_GPIOZ(1):
-		case STM32MP1_SHRES_GPIOZ(2):
-		case STM32MP1_SHRES_GPIOZ(3):
-		case STM32MP1_SHRES_GPIOZ(4):
-		case STM32MP1_SHRES_GPIOZ(5):
-		case STM32MP1_SHRES_GPIOZ(6):
-		case STM32MP1_SHRES_GPIOZ(7):
-			stm32mp_register_clock_parents_secure(GPIOZ);
-			break;
 		case STM32MP1_SHRES_IWDG1:
 			stm32mp_register_clock_parents_secure(IWDG1);
 			break;
@@ -352,80 +319,6 @@ void stm32mp_register_non_secure_periph_iomem(vaddr_t base)
 	register_periph_iomem(base, SHRES_NON_SECURE);
 }
 
-/* Register GPIO resource */
-void stm32mp_register_secure_gpio(unsigned int bank, unsigned int pin)
-{
-	switch (bank) {
-	case GPIO_BANK_Z:
-		assert(pin < get_gpioz_nbpin());
-		register_periph(STM32MP1_SHRES_GPIOZ(pin), SHRES_SECURE);
-		break;
-	default:
-		EMSG("GPIO bank %u cannot be secured", bank);
-		panic();
-	}
-}
-
-void stm32mp_register_non_secure_gpio(unsigned int bank, unsigned int pin)
-{
-	switch (bank) {
-	case GPIO_BANK_Z:
-		assert(pin < get_gpioz_nbpin());
-		register_periph(STM32MP1_SHRES_GPIOZ(pin), SHRES_NON_SECURE);
-		break;
-	default:
-		break;
-	}
-}
-
-void stm32mp_register_secure_pinctrl(struct pinctrl_state *pinctrl)
-{
-	unsigned int *bank = NULL;
-	unsigned int *pin = NULL;
-	size_t count = 0;
-	size_t n = 0;
-
-	stm32_gpio_pinctrl_bank_pin(pinctrl, NULL, NULL, &count);
-	if (!count)
-		return;
-
-	bank = calloc(count, sizeof(*bank));
-	pin = calloc(count, sizeof(*pin));
-	if (!bank || !pin)
-		panic();
-
-	stm32_gpio_pinctrl_bank_pin(pinctrl, bank, pin, &count);
-	for (n = 0; n < count; n++)
-		stm32mp_register_secure_gpio(bank[n], pin[n]);
-
-	free(bank);
-	free(pin);
-}
-
-void stm32mp_register_non_secure_pinctrl(struct pinctrl_state *pinctrl)
-{
-	unsigned int *bank = NULL;
-	unsigned int *pin = NULL;
-	size_t count = 0;
-	size_t n = 0;
-
-	stm32_gpio_pinctrl_bank_pin(pinctrl, NULL, NULL, &count);
-	if (!count)
-		return;
-
-	bank = calloc(count, sizeof(*bank));
-	pin = calloc(count, sizeof(*pin));
-	if (!bank || !pin)
-		panic();
-
-	stm32_gpio_pinctrl_bank_pin(pinctrl, bank, pin, &count);
-	for (n = 0; n < count; n++)
-		stm32mp_register_non_secure_gpio(bank[n], pin[n]);
-
-	free(bank);
-	free(pin);
-}
-
 static void lock_registering(void)
 {
 	registering_locked = true;
@@ -438,40 +331,6 @@ bool stm32mp_periph_is_secure(enum stm32mp_shres id)
 	return shres_state[id] == SHRES_SECURE;
 }
 
-bool stm32mp_gpio_bank_is_non_secure(unsigned int bank)
-{
-	unsigned int not_secure = 0;
-	unsigned int pin = 0;
-
-	lock_registering();
-
-	if (bank != GPIO_BANK_Z)
-		return true;
-
-	for (pin = 0; pin < get_gpioz_nbpin(); pin++)
-		if (!stm32mp_periph_is_secure(STM32MP1_SHRES_GPIOZ(pin)))
-			not_secure++;
-
-	return not_secure > 0 && not_secure == get_gpioz_nbpin();
-}
-
-bool stm32mp_gpio_bank_is_secure(unsigned int bank)
-{
-	unsigned int secure = 0;
-	unsigned int pin = 0;
-
-	lock_registering();
-
-	if (bank != GPIO_BANK_Z)
-		return false;
-
-	for (pin = 0; pin < get_gpioz_nbpin(); pin++)
-		if (stm32mp_periph_is_secure(STM32MP1_SHRES_GPIOZ(pin)))
-			secure++;
-
-	return secure > 0 && secure == get_gpioz_nbpin();
-}
-
 static bool mckprot_resource(enum stm32mp_shres id)
 {
 	switch (id) {
diff --git a/core/arch/arm/plat-stm32mp1/stm32_util.h b/core/arch/arm/plat-stm32mp1/stm32_util.h
index 7b35ef5be..65a9b06f9 100644
--- a/core/arch/arm/plat-stm32mp1/stm32_util.h
+++ b/core/arch/arm/plat-stm32mp1/stm32_util.h
@@ -208,46 +208,9 @@ void stm32mp_register_secure_periph_iomem(vaddr_t base);
  */
 void stm32mp_register_non_secure_periph_iomem(vaddr_t base);
 
-/*
- * Register GPIO resource as a secure peripheral
- * @bank: Bank of the target GPIO
- * @pin: Bit position of the target GPIO in the bank
- */
-void stm32mp_register_secure_gpio(unsigned int bank, unsigned int pin);
-
-/*
- * Register GPIO resource as a non-secure peripheral
- * @bank: Bank of the target GPIO
- * @pin: Bit position of the target GPIO in the bank
- */
-void stm32mp_register_non_secure_gpio(unsigned int bank, unsigned int pin);
-
-/*
- * Register pin resource of a pin control state as a secure peripheral
- * @bank: Bank of the target GPIO
- * @pin: Bit position of the target GPIO in the bank
- */
-void stm32mp_register_secure_pinctrl(struct pinctrl_state *pinctrl);
-
-/*
- * Register pin resource of a pin control state as a non-secure peripheral
- * @bank: Bank of the target GPIO
- * @pin: Bit position of the target GPIO in the bank
- */
-void stm32mp_register_non_secure_pinctrl(struct pinctrl_state *pinctrl);
-
 /* Return true if and only if resource @id is registered as secure */
 bool stm32mp_periph_is_secure(enum stm32mp_shres id);
 
-/* Return true if and only if GPIO bank @bank is registered as secure */
-bool stm32mp_gpio_bank_is_secure(unsigned int bank);
-
-/* Return true if and only if GPIO bank @bank is registered as non-secure */
-bool stm32mp_gpio_bank_is_non_secure(unsigned int bank);
-
-/* Register number of pins in the GPIOZ bank */
-void stm32mp_register_gpioz_pin_count(size_t count);
-
 #else /* CFG_STM32MP1_SHARED_RESOURCES */
 
 static inline void stm32mp_register_secure_periph(enum stm32mp_shres id
@@ -269,44 +232,10 @@ static inline void stm32mp_register_non_secure_periph_iomem(vaddr_t base
 {
 }
 
-static inline void stm32mp_register_secure_gpio(unsigned int bank __unused,
-						unsigned int pin __unused)
-{
-}
-
-static inline void stm32mp_register_non_secure_gpio(unsigned int bank __unused,
-						    unsigned int pin __unused)
-{
-}
-
-static inline void
-stm32mp_register_secure_pinctrl(struct pinctrl_state *pinctrl __unused)
-{
-}
-
-static inline void
-stm32mp_register_non_secure_pinctrl(struct pinctrl_state *pinctrl __unused)
-{
-}
-
 static inline bool stm32mp_periph_is_secure(enum stm32mp_shres id __unused)
 {
 	return true;
 }
-
-static inline bool stm32mp_gpio_bank_is_secure(unsigned int bank __unused)
-{
-	return true;
-}
-
-static inline bool stm32mp_gpio_bank_is_non_secure(unsigned int bank __unused)
-{
-	return false;
-}
-
-static inline void stm32mp_register_gpioz_pin_count(size_t count __unused)
-{
-}
 #endif /* CFG_STM32MP1_SHARED_RESOURCES */
 
 void __noreturn do_reset(const char *str);
diff --git a/core/drivers/stm32_gpio.c b/core/drivers/stm32_gpio.c
index f0c2aa82d..2a27b013b 100644
--- a/core/drivers/stm32_gpio.c
+++ b/core/drivers/stm32_gpio.c
@@ -1212,9 +1212,6 @@ static TEE_Result dt_stm32_gpio_bank(const void *fdt, int node,
 		bank->base = io_pa_or_va_nsec(&pa_va, blen);
 	}
 
-	if (compat->gpioz)
-		stm32mp_register_gpioz_pin_count(bank->ngpios);
-
 	*out_bank = bank;
 
 	return TEE_SUCCESS;
-- 
2.47.2

